src/HW_models/NRF_RTC.c

```diff
@@ -39,90 +39,142 @@
 #define N_CC 4
 
 #define RTC_COUNTER_MASK 0xFFFFFF /*24 bits*/
+#define RTC_TRIGGER_OVERFLOW_COUNTER_VALUE 0xFFFFF0
+
+#define SUB_US_BITS 9 // Bits representing sub-microsecond units
+
 NRF_RTC_Type NRF_RTC_regs[N_RTC];
 
 static bool RTC_Running[N_RTC] = {false};
 static uint32_t RTC_INTEN[N_RTC] = {0};
 
 bs_time_t Timer_RTC = TIME_NEVER;
 static bs_time_t cc_timers[N_RTC][N_CC] = {{TIME_NEVER}}; //when each CC will match (in microseconds)
+static bs_time_t overflow_timer[N_RTC] = {TIME_NEVER}; //when the timer will overflow (in microseconds)
+
+static bs_time_t overflow_timer_sub_us[N_RTC] = {TIME_NEVER}; //when the timer will overflow (in sub-microsecond units)
+static bs_time_t RTC_counter_startT_sub_us[N_RTC] = {TIME_NEVER}; //Time when the counter was "started" (really the time that would correspond to COUNTER = 0)
+static bs_time_t RTC_counter_startT_negative_sub_us[N_RTC] = {0};
 
-static bs_time_t RTC_counter_startT[N_RTC] = {TIME_NEVER}; //Time when the counter was "started" (really the time that would correspond to COUNTER = 0)
 static uint32_t counter[N_RTC] = {0}; //Internal counter value when the counter was stopped
 
-static bs_time_t first_lf_tick_time = 0;
+static bs_time_t first_lf_tick_time_sub_us = 0;
 
-void nrf_rtc_init() {
-  memset(NRF_RTC_regs, 0, sizeof(NRF_RTC_regs));
-  for (int i = 0; i < N_RTC ; i++) {
-    RTC_Running[i] = false;
-    RTC_INTEN[i] = 0;
-    counter[i] = 0;
-    RTC_counter_startT[i] = TIME_NEVER;
-    for (int j  = 0 ; j < N_CC ; j++) {
-      cc_timers[i][j] = TIME_NEVER;
-    }
-  }
-  Timer_RTC = TIME_NEVER;
-}
 
-void nrf_rtc_clean_up() {
+static bs_time_t sub_us_time_to_us_time(bs_time_t sub_us_time);
+static bs_time_t us_time_to_sub_us_time(bs_time_t us_time);
 
-}
+/**
+ * Convert a time delta in sub-microseconds units to the equivalent time in microseconds.
+ * The value is always rounded UP. This is because otherwise the events would be registered
+ * in a time in microseconds before the event actually occured. This would lead to many imprecise
+ * event timings for example if the timing of an event would be calculated base on the last LF
+ * clock tick (which happens for example when triggering the CLEAR or TRIGGER_OVERFLOW tasks)
+ */
+static bs_time_t sub_us_time_to_us_time(bs_time_t sub_us_time)
+{
+  bs_time_t us_time = sub_us_time >> SUB_US_BITS;
 
-void nrf_rtc_notify_first_lf_tick() {
-  first_lf_tick_time = tm_get_hw_time();
-  bs_trace_raw_time(9, "RTC: First lf tick\n");
+  if(sub_us_time % (1U << SUB_US_BITS) != 0) //rounding up
+  {
+    us_time += 1;
+  }
+
+  return us_time;
 }
 
+/**
+ * Convert a time delta in microseconds to the equivalent time in sub-microseconds units
+ * (crop if the time in microseconds is not an integer)
```

Comment:
"(crop if the time in microseconds is not an integer)" what does this mean?

Reply:
This comment must have been somehow left during copying and is not relevant anymore, deleted.